File: app/src/main/java/com/blackmirror/dongda/adapter/HomeScienceAdapter.java
package com.blackmirror.dongda.adapter;

import android.content.Context;
import android.support.v7.widget.RecyclerView;
import android.text.TextUtils;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;

import com.blackmirror.dongda.R;
import com.blackmirror.dongda.Tools.OSSUtils;
import com.blackmirror.dongda.model.serverbean.HomeInfoServerBean;
import com.facebook.drawee.view.SimpleDraweeView;

import java.util.List;

public class HomeScienceAdapter extends RecyclerView.Adapter<HomeScienceAdapter.HomeScienceViewHolder> {


    HomeInfoServerBean.ResultBean.HomepageServicesBean bean;
    protected Context context;
    private OnItemClickListener listener;

    public void setOnItemClickListener(OnItemClickListener listener) {
        this.listener = listener;
    }

    public HomeScienceAdapter(Context context,  HomeInfoServerBean.ResultBean.HomepageServicesBean bean) {
        this.context = context;
        this.bean = bean;
    }

    @Override
    public HomeScienceViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        View view = LayoutInflater.from(parent.getContext()).inflate(R.layout.rv_item_home_science,parent,false);
        return new HomeScienceViewHolder(view);
    }

    @Override
    public void onBindViewHolder(HomeScienceViewHolder holder, int position) {


        HomeInfoServerBean.ResultBean.HomepageServicesBean.ServicesBean servicesBean = this.bean.services.get(position);

        String url= OSSUtils.getSignedUrl(servicesBean.service_image,30*60);
        holder.sv_item_science_photo.setImageURI(url);

        /*try {
            String url = GetOSSClient.INSTANCE().oss.presignConstrainedObjectURL("bm-dongda", servicesBean.service_image+".jpg", 30 * 60);
            Log.d("xcx", "url: "+url);
            holder.sv_item_science_photo.setImageURI(url);
        } catch (ClientException e) {
            e.printStackTrace();
        }*/

        if (servicesBean.is_collected){
            holder.iv_item_science_like.setBackgroundResource(R.drawable.like_selected);
        }else {
            holder.iv_item_science_like.setBackgroundResource(R.drawable.home_art_like);
        }

        if (TextUtils.isEmpty(bean.services.get(position).service_tags.get(0))){
            holder.tv_item_science_name.setVisibility(View.GONE);
        }else {
            holder.tv_item_science_name.setVisibility(View.VISIBLE);
            holder.tv_item_science_name.setText(bean.services.get(position).service_tags.get(0));
        }

        StringBuilder sb = new StringBuilder();
        sb.append(servicesBean.brand_name)
                .append("的")
                .append(servicesBean.service_leaf)
                .append(servicesBean.category);
        holder.tv_item_science_detail.setText(sb.toString());
        holder.tv_item_science_location.setText(servicesBean.address.substring(0,servicesBean.address.indexOf("区")+1));

        initListener(holder, position, servicesBean);

    }

    @Override
    public void onBindViewHolder(HomeScienceViewHolder holder, int position, List<Object> payloads) {

        if (payloads.isEmpty()) {
            onBindViewHolder(holder, position);
        } else {
            boolean isLike= (boolean) payloads.get(0);
            bean.services.get(position).is_collected= isLike;
            if (isLike){
                holder.iv_item_science_like.setBackgroundResource(R.drawable.like_selected);
            }else {
                holder.iv_item_science_like.setBackgroundResource(R.drawable.home_art_like);
            }
        }
    }

    private void initListener(final HomeScienceViewHolder holder, int position, final HomeInfoServerBean.ResultBean.HomepageServicesBean.ServicesBean servicesBean) {

        holder.itemView.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                if (listener != null) {
                    int pos = holder.getLayoutPosition();
                    listener.onScienceItemClick(holder.iv_item_science_like, pos, servicesBean.service_id);
                }
            }
        });

        holder.iv_item_science_like.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                if (listener != null) {
                    int pos = holder.getLayoutPosition();
                    listener.onScienceLikeClick(holder.iv_item_science_like, pos, servicesBean);
                }
            }
        });
    }

    @Override
    public int getItemCount() {
        if (bean.services != null) {
            return bean.services.size();
        }
        return 0;
    }

    public void setRefreshData(List<HomeInfoServerBean.ResultBean.HomepageServicesBean.ServicesBean> list){
        bean.services.clear();
        bean.services.addAll(list);
        notifyDataSetChanged();
    }

    public static class HomeScienceViewHolder extends RecyclerView.ViewHolder {

        public SimpleDraweeView sv_item_science_photo;
        public ImageView iv_item_science_like;
        public TextView tv_item_science_name;
        public TextView tv_item_science_detail;
        public TextView tv_item_science_location;

        public HomeScienceViewHolder(View itemView) {
            super(itemView);
            sv_item_science_photo = itemView.findViewById(R.id.sv_item_science_photo);
            iv_item_science_like = itemView.findViewById(R.id.iv_item_science_like);
            tv_item_science_name = itemView.findViewById(R.id.tv_item_science_name);
            tv_item_science_detail = itemView.findViewById(R.id.tv_item_science_detail);
            tv_item_science_location = itemView.findViewById(R.id.tv_item_science_location);

        }
    }

    public interface OnItemClickListener {
        void onScienceLikeClick(View view, int position, HomeInfoServerBean.ResultBean
                .HomepageServicesBean.ServicesBean servicesBean);

        void onScienceItemClick(View view, int position, String service_id);

    }

}

File: app/src/main/java/com/blackmirror/dongda/adapter/HomeScienceAdapter.kt
package com.blackmirror.dongda.adapter

import android.content.Context
import android.support.v7.widget.RecyclerView
import android.text.TextUtils
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.TextView
import com.blackmirror.dongda.R
import com.blackmirror.dongda.kdomain.model.HomepageDomainBean
import com.blackmirror.dongda.utils.OSSUtils
import com.facebook.drawee.view.SimpleDraweeView

class HomeScienceAdapter(protected var context: Context, internal var bean: HomepageDomainBean.HomepageServicesBean) : RecyclerView.Adapter<HomeScienceAdapter.HomeScienceViewHolder>() {

    private var item: ((View, Int, String?) -> Unit)? = null
    private var like: ((View, Int, HomepageDomainBean.HomepageServicesBean.ServicesBean) -> Unit)? = null

    fun setOnItemClickListener(it: ((View, Int, String?) -> Unit)? = null, li: ((View, Int, HomepageDomainBean.HomepageServicesBean.ServicesBean) -> Unit)? = null) {
        item = it
        like = li
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): HomeScienceViewHolder {
        val view = LayoutInflater.from(parent.context).inflate(R.layout.rv_item_home_science, parent, false)
        return HomeScienceViewHolder(view)
    }

    override fun onBindViewHolder(holder: HomeScienceViewHolder, position: Int) {


        val servicesBean = this.bean.services!![position]

        val url = OSSUtils.getSignedUrl(servicesBean.service_image, (30 * 60).toLong())
        holder.sv_item_science_photo.setImageURI(url)


        if (servicesBean.is_collected) {
            holder.iv_item_science_like.setBackgroundResource(R.drawable.like_selected)
        } else {
            holder.iv_item_science_like.setBackgroundResource(R.drawable.home_art_like)
        }

        if (TextUtils.isEmpty(bean.services!![position].operation!![0])) {
            holder.tv_item_science_name.visibility = View.GONE
        } else {
            holder.tv_item_science_name.visibility = View.VISIBLE
            holder.tv_item_science_name.text = bean.services!![position].operation!![0]
        }

        val sb = StringBuilder()
        sb.append(servicesBean.brand_name)
                .append(context.getString(R.string.str_de))
                .append(servicesBean.service_leaf)
                .append(servicesBean.category)
        holder.tv_item_science_detail.text = sb.toString()
        holder.tv_item_science_location.text = servicesBean.address!!.substring(0, servicesBean.address!!.indexOf(context.getString(R.string.str_qu)) + 1)

        initListener(holder, position, servicesBean)

    }

    override fun onBindViewHolder(holder: HomeScienceViewHolder, position: Int, payloads: List<Any>?) {

        if (payloads!!.isEmpty()) {
            onBindViewHolder(holder, position)
        } else {
            val isLike = payloads[0] as Boolean
            bean.services!![position].is_collected = isLike
            if (isLike) {
                holder.iv_item_science_like.setBackgroundResource(R.drawable.like_selected)
            } else {
                holder.iv_item_science_like.setBackgroundResource(R.drawable.home_art_like)
            }
        }
    }

    private fun initListener(holder: HomeScienceViewHolder, position: Int, servicesBean: HomepageDomainBean.HomepageServicesBean.ServicesBean) {

        holder.itemView.setOnClickListener {
            item?.invoke(holder.iv_item_science_like, holder.layoutPosition, servicesBean.service_id)
        }

        holder.iv_item_science_like.setOnClickListener {
            like?.invoke(holder.iv_item_science_like, holder.layoutPosition, servicesBean)
        }
    }

    override fun getItemCount(): Int {

        return if (bean.services != null) {
            bean.services!!.size
        } else 0
    }

    fun setRefreshData(list: MutableList<HomepageDomainBean.HomepageServicesBean.ServicesBean>?) {
        list?.apply {
            bean.services?.clear()
            bean.services?.addAll(this)
        }
        notifyDataSetChanged()
    }

    class HomeScienceViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {

        var sv_item_science_photo: SimpleDraweeView
        var iv_item_science_like: ImageView
        var tv_item_science_name: TextView
        var tv_item_science_detail: TextView
        var tv_item_science_location: TextView

        init {
            sv_item_science_photo = itemView.findViewById(R.id.sv_item_science_photo)
            iv_item_science_like = itemView.findViewById(R.id.iv_item_science_like)
            tv_item_science_name = itemView.findViewById(R.id.tv_item_science_name)
            tv_item_science_detail = itemView.findViewById(R.id.tv_item_science_detail)
            tv_item_science_location = itemView.findViewById(R.id.tv_item_science_location)

        }
    }
}
